alphacsc/_d_solver.py

```diff
@@ -124,44 +222,16 @@ def init_dictionary(self, X, n_atoms, n_times_atom, D_init=None,
             The initial atoms to learn from the data.
         """
 
-        return init_dictionary(
-            X, n_atoms, n_times_atom, D_init=D_init,
-            rank1=self.rank1, uv_constraint=self.uv_constraint,
-            D_init_params=D_init_params, random_state=self.rng,
-            window=self.window
-        )
-
-    def update_D(self, z_encoder):
-        """Learn d's in time domain.
-
-        Parameters
-        ----------
-        z_encoder: BaseZEncoder
-            ZEncoder object.
+        D_hat = self.init_strategy.initialize(X)
 
-        Returns
-        -------
-        D_hat : array, shape (n_atoms, n_channels + n_times_atom) or
-                             (n_atoms, n_channels, n_times_atom)
-            The atoms to learn from the data.
-        """
-
-        self._init_windower(z_encoder)
-
-        D_hat0 = self.windower.dewindow(z_encoder.D_hat)
-
-        D_hat, pobj = fista(
-            self._get_objective(z_encoder), self._get_grad(z_encoder),
-            self._get_prox(z_encoder), None, D_hat0, self.max_iter,
-            momentum=self.momentum, eps=self.eps, adaptive_step_size=True,
-            name=self.name, debug=self.debug, verbose=self.verbose
-        )
+        # if D_init is not an ndarray
+        # when D_init is an ndarray, the strategy has D_init attribute
+        if not hasattr(self.init_strategy, 'D_init'):
```

Comment:
Maybe better to check this with the type of `IdentityStrategy`?
```suggestion
        # Only use window on D_hat if D_init is not an ndarray
        if not isinstance(self.init_strategy, IdentityStrategy):
```

Reply:
I think it would be better to avoid passing concrete strategy class name to `d_solver`. It does not have to know about class types. 

As a solution I've set `D_init` as class variable in `d_solver`. And I check as in original code, if it is of type `ndarray`. Another option would be to set it in `strategy` class.